### PR TITLE
JS: generalize js/incomplete-multi-character-sanitization

### PR DESCRIPTION
**Pull request by esbena**
This is a revival of some generalization for the `js/incomplete-multi-character-sanitization` query, the new thing is that regular expressions are treated with much more generality.

The [new results](https://lgtm.com/query/5208022070019174423/) are for replacements that may behave suprisingly (TPs), yet they do not appear to be for sanitizers (so they are not security TPs). The previous version of this query also has this problem.

@erik-krogh, can you do a performance sanity check (I no longer have access to the workers)?

### CODE REVIEW

**Comment by esbena** [comment]
(it looks like there's a performance problem that I need to sort out, sorry for the early evaluation request)

**Comment by esbena** [comment]
I have force-pushed another faster rewrite, the results have improved a tiny bit in theory but not in practice (https://lgtm.com/query/271231760340743802/).

The crucial optimization is [here](https://github.com/github/codeql/pull/4928/files#diff-e8b7e7e84899fe415604b9a11df8ab1a5efc4bced79918f0908df83860e339edR33): Since we statically know the strings we care about, the search for interesting regexps can be made very fast.

File: javascript/ql/src/Security/CWE-116/IncompleteMultiCharacterSanitization.ql [hunk 1]
```diff
@@ -12,69 +12,153 @@
...
+    kind = "path injection" and
+    // upwards navigation
+    prefix = ["/..", "../"] and
+    not t.getSuccessor*().getAMatchedString().regexpMatch("(?i).*[a-z0-9_-]+.*") // explicit path name mentions make this an unlikely sanitizer
+    or
+    kind = "HTML element injection" and
+    (
+      // comments
+      prefix = "<!--" and
+      not t.getSuccessor*().getAMatchedString().regexpMatch("(?i).*[a-z0-9_]+.*") // explicit comment content mentions make this an unlikely sanitizer
```
Comment (asgerf) [inline javascript/ql/src/Security/CWE-116/IncompleteMultiCharacterSanitization.ql:98]:
```suggestion
      not t.getSuccessor*().getAMatchedString().regexpMatch("(?is).*[a-z0-9_].*") // explicit comment content mentions make this an unlikely sanitizer
```

**Review by asgerf — COMMENTED** [review]
Some of the alerts read:

> This string may still contain a substring that starts matching at `../`

I'm not sure exactly what is meant by this, but it doesn't seem right. In the attack string `....//`, the resulting `../` is not at the same index as the removed `../` substring; it starts two characters earlier.

Is there anything wrong with just saying "the string may still contain `../`"?

**Comment by erik-krogh** [comment]
> I have force-pushed another faster rewrite, the results have improved a tiny bit in theory but not in practice (https://lgtm.com/query/271231760340743802/).

[An evaluation of that](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1610389755006).

**Comment by esbena** [comment]
Thanks. All comments addressed.

> This string may still contain a substring that starts matching at ../
 
The intention was to tell the programmer where the regexp start matching `../`: it is a link that points to the regexp term that matches at least the first `.` of `../`. But the suggested simpler formulation is better and less ambiguous.

**Comment by esbena** [comment]
I think this is ready for approval.

**Review by asgerf — APPROVED** [review]

**Pull request merged**